cpp/src/cuda_event.cpp

```diff
@@ -0,0 +1,60 @@
+/**
+ * SPDX-FileCopyrightText: Copyright (c) 2025, NVIDIA CORPORATION & AFFILIATES.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <atomic>
+
+#include <cuda_runtime_api.h>
+
+#include <rmm/cuda_stream_view.hpp>
+
+#include <rapidsmpf/cuda_event.hpp>
+
+namespace rapidsmpf {
+
+Event::Event() {
+    RAPIDSMPF_CUDA_TRY(cudaEventCreateWithFlags(&event_, cudaEventDisableTiming));
+}
+
+Event::~Event() {
+    cudaEventDestroy(event_);
+}
+
+void Event::record(rmm::cuda_stream_view stream) const {
+    if (!query()) {
+        RAPIDSMPF_CUDA_TRY(cudaEventRecord(event_, stream.value()));
+    }
+}
+
+void Event::add_as_dependency_to(rmm::cuda_stream_view stream) const {
+    if (!query()) {
+        RAPIDSMPF_CUDA_TRY(
+            cudaStreamWaitEvent(stream.value(), event_, cudaEventWaitDefault)
+        );
+    }
+}
```

Comment:
This is preparing for if we start using an internal stream inside the shuffler that is distinct from the user stream. In that case, when we copy a user buffer on the internal stream we must add a dependency on the work on the user's stream.